File: Data-Structure/Queue/Double Ended Queue/Double Ended Deletion Queue/DEDQ.java
import java.util.Scanner;

class DoubleEndedDeletionQueue {
    // Queue Initialization
    private final int MAX = 5;
    private int[] queue = new int[MAX];
    private int BACK=-1, FRONT=-1;

    // OVERFLOW CONDITION
    public boolean OVERFLOW() {
        if (BACK == MAX-1) {
            return true;
        } else {
            return false;
        }
    }

    // UNDERFLOW CONDITION
    public boolean UNDERFLOW() {
        if (BACK==-1) {
            return true;
        } else {
            return false;
        }
    }

    // insert into queue
    public boolean insert(int item) {
        if(!OVERFLOW()) {
            if (BACK == -1)
                FRONT = 0;
            queue[++BACK] = item;
            return true;
        } else {
            return false;
        }
    }

    // delete from queue BACK
    public int delete_back() {
        if (!UNDERFLOW()) {
            int temp = queue[FRONT];
            if (FRONT == BACK)
                FRONT = BACK = -1;
            else
                BACK--;
            return temp;
        } else {
            return -1;
        }
    }

    // delete from queue FRONT
    public int delete_front() {
        if (!UNDERFLOW()) {
            int temp = queue[FRONT];
            if (FRONT == BACK)
                FRONT = BACK = -1;
            else
                FRONT++;
            return temp;
        } else {
            return -1;
        }
    }

    // Display item in queue
    public void display() {
        if(!UNDERFLOW()) {
        for (int i=FRONT; i<=BACK; i++) {
                System.out.print(queue[i]+"\t");
            }
            System.out.println();
        } else {
            System.out.println("-> UNDERFLOW <-");
        }
    }
}

public class DEDQ {
    // DIRVER CODE
    public static void main(String[] args) {
        DoubleEndedDeletionQueue queue = new DoubleEndedDeletionQueue();

        Scanner sc = new Scanner(System.in);
        int switch_var;		    // Switch Variable
        boolean exit = false;	// Exit condition
        int temp;			    // Temporary variable
    
        while (!exit) {
            // Menu
            System.out.print("\nEnter \n\t1. INSERT \n\t2. DELETE BACK \n\t3. DELETE FRONT \n\t4. DISPLAY \n\t5. EXIT \n-> ");
            switch_var = sc.nextInt();
    
            switch (switch_var) {
                // insert
                case 1:
                    System.out.print("Enter a value to insert : ");
                    temp = sc.nextInt();
                    if (queue.insert(temp)) {  // If not oveflowed
                        System.out.println(temp+" is inserted into queue.");
                    } else {
                        System.out.println("-> OVERFLOW <-");
                    }
                    break;
                // delete BACK
                case 2:
                    temp = queue.delete_back();
                    if (temp == -1) {  // If underflowed
                        System.out.println("-> UNDERFLOW <-");                        
                    } else {
                        System.out.println("The deleted value is "+temp+".");
                    }
                    break;
                // delete FRONT
                case 3:
                    temp = queue.delete_front();
                    if (temp == -1) {  // If underflowed
                        System.out.println("-> UNDERFLOW <-");
                    } else {
                        System.out.println("The deleted value is "+temp+".");
                    }
                    break;
                // DISPLAY
                case 4:
                    queue.display();
                    break;
                // EXIT
                case 5:
                    exit = true;
                    break;
                // INVALID
                default:
                    System.out.println("Enter a valid option");
                    break;
            }
        }
        sc.close();
    }
}

File: Data-Structure/Queue/Double Ended Queue/Double Ended Deletion Queue/DEDQ.py
class Queue():
    # Queue Initialization
    def __init__(self):
        self.MAX = 5
        self.queue = []

    # OVERFLOW CONDITION
    def OVERFLOW(self):
        if len(self.queue) == self.MAX:
            return True
        else:
            return False

    # UNDERFLOW CONDITION
    def UNDERFLOW(self):
        if len(self.queue) == 0:
            return True
        else:
            return False
    
    # Insert into queue
    def insert(self, item):
        if not self.OVERFLOW():
            self.queue.append(item)
            return True
        else:
            return False
    
    # Delete from queue FRONT
    def delete_front(self):
        if not self.UNDERFLOW():
            return self.queue.pop(0)
        else:
            return False

    # Delete from queue BACK
    def delete_back(self):
        if not self.UNDERFLOW():
            return self.queue.pop()
        else:
            return False
    
    def display(self):
        if not self.UNDERFLOW():
            print("   ".join([str(i) for i in self.queue]))
        else:
            print("-> UNDERFLOW <-")

menu = '''
Enter
    1. INSERT
    2. DELETE FRONT
    3. DELETE BACK
    4. DISPAY
    5. EXIT
-> '''

if __name__ == "__main__":
    obj = Queue()
    exit = False
    while not exit:
        switch_var = int(input(menu))
        # INSERT
        if switch_var == 1:
            temp = int(input("Enter the number : "))
            if obj.insert(temp):
                print(f"{temp} is inserted into queue.")
            else:
                print("-> OVERFLOW <-")
        # DELETE BACK
        elif switch_var == 2:
            temp = obj.delete_front()
            if temp:
                print(f"{temp} is deleted from queue.")
            else:
                print("-> UNDERFLOW <-")
        # DELETE BACK
        elif switch_var == 3:
            temp = obj.delete_back()
            if temp:
                print(f"{temp} is deleted from queue.")
            else:
                print("-> UNDERFLOW <-")
        # DISPLAY
        elif switch_var == 4:
            obj.display()
        # EXIT
        elif switch_var == 5:
            exit = True
        # INVALID
        else:
            print("Please enter a valid statement.")
